### PR TITLE
fix(1412): allow week/month/year aggregation

### PR DESCRIPTION
**Pull request by d2lam**
This will allow aggregation option like `day/week/month/year`

Related: https://github.com/screwdriver-cd/screwdriver/issues/1412

### CODE REVIEW

File: lib/job.js [hunk 1]
```diff
@@ -275,7 +275,8 @@ class Job extends BaseModel {
         }
 
         options.paginate.page = 1;
-        const allBuilds = await getAllRecords.apply(this, ['getBuilds', options, [[]]]);
+        const allBuilds = await getAllRecords.apply(this,
+            ['getBuilds', config.aggregate, options, [[]]]);
```
Comment (DekusDenial) [inline lib/job.js]: wondering if u know all the args, why not use `.call`?

File: lib/helper.js [hunk 2]
```diff
@@ -54,13 +54,18 @@ function parseTemplateConfigName(config) {
...
+ * @param  {String}       aggregate   Aggregate data by day/month/week/year
  * @param  {Object}       opts        opts to pass into scan function
  * @param  {Array}        resultArray Result so far
  * @param  {String}       date        Current date for aggregation
  * @param  {Number}       index       Current index (each index is a different date)
  * @return {Array}                    Array where each element is an array of records belong to the same date
  */
-async function getAllRecords(funcName, opts, resultArray, date, index) {
+async function getAllRecords(funcName, aggregate, opts, resultArray, date, index) {
+    if (!['day', 'week', 'month', 'year'].includes(aggregate)) {
```
Comment (DekusDenial) [inline lib/helper.js]: https://github.com/screwdriver-cd/screwdriver/pull/1582#discussion_r274465814

File: lib/pipeline.js [hunk 3]
```diff
@@ -1292,11 +1292,12 @@ class PipelineModel extends BaseModel {
 
         options.paginate.page = 1;
 
-        const allEvents = await getAllRecords.apply(this, ['getEvents', options, [[]]]);
+        const allEvents = await getAllRecords.apply(this,
+            ['getEvents', config.aggregate, options, [[]]]);
```
Comment (DekusDenial) [inline lib/pipeline.js]: same here

File: lib/job.js (hunk 1 above)
Comment (d2lam) [reply lib/job.js]: I guess I can use either. Are there any difference performance difference?

Comment (DekusDenial) [reply lib/job.js]: Technical answer: yes
Is it a lot? Depends on number of args
Is this blocking? No, hence this is just a comment ;-P

Comment (d2lam) [reply lib/job.js]: Just curious, which one has better performance?

Comment (DekusDenial) [reply lib/job.js]: Call is bit faster than apply

File: lib/job.js [hunk 4]
```diff
@@ -13,19 +13,19 @@ const MAX_COUNT = 1000;
 /**
  * Find metrics and step metrics related to the build
  * @method findMetrics
- * @param  {Object}    build      build object
- * @param  {String}    stepName   only include this step
- * @param  {Boolean}   aggregate  Whether to include step metrics
- * @return {Array}                Array of metrics
+ * @param  {Object}    build              build object
+ * @param  {String}    stepName           only include this step
+ * @param  {Boolean}   aggregateInterval  Include step metrics if no aggregation
```
Comment (tkyi) [inline lib/job.js]: ~Boolean~ String

File: lib/job.js [hunk 5]
```diff
@@ -249,11 +249,11 @@ class Job extends BaseModel {
...
-     * @param  {Object}   [config]              Configuration object
-     * @param  {String}   [config.startTime]    Look at builds created after this startTime
-     * @param  {String}   [config.endTime]      Look at builds created before this endTime
-     * @param  {String}   [config.stepName]     Only include this step
-     * @param  {Boolean}  [config.aggregate]    Whether to aggregate the data
+     * @param  {Object}   [config]                    Configuration object
+     * @param  {String}   [config.startTime]          Look at builds created after this startTime
+     * @param  {String}   [config.endTime]            Look at builds created before this endTime
+     * @param  {String}   [config.stepName]           Only include this step
+     * @param  {Boolean}  [config.aggregateInterval]  Whether to aggregateInterval the data
```
Comment (tkyi) [inline lib/job.js]: ~Boolean~ String

**Review by tkyi — APPROVED** [review]

File: lib/helper.js [hunk 6]
```diff
@@ -53,14 +53,19 @@ function parseTemplateConfigName(config) {
...
+ * @param  {String}       aggregateInterval   Aggregate data by day/month/week/year
+ * @param  {Object}       opts                opts to pass into scan function
+ * @param  {Array}        resultArray         Result so far
+ * @param  {String}       date                Current date for aggregation
+ * @param  {Number}       index               Current index (each index is a different date)
+ * @return {Array}                            Array where each element is an array of records belong to the same date
  */
-async function getAllRecords(funcName, opts, resultArray, date, index) {
+async function getAllRecords(funcName, aggregateInterval, opts, resultArray, date, index) {
+    if (!['day', 'week', 'month', 'year'].includes(aggregateInterval)) {
```
Comment (minzcmu) [inline lib/helper.js]: do we still need to check here? seeing API has this check already for the query param

**Review by minzcmu — APPROVED** [review]
small comment, 👍

File: lib/helper.js (hunk 6 above)
Comment (d2lam) [reply lib/helper.js]: probably don't need. I wrote this before the API, so that's why I had this check. But API already checks so I can remove it 👍

**Pull request merged**